### PR TITLE
Add FileSessionConfAdvisor to manage session level configuration

### PR DESCRIPTION
**Pull request by ychris78**
### _Why are the changes needed?_

When a kyuubi service connections to multiple engine clusters, we can only explicitly specify relevant parameters after the URL, for example: 
`beeline -u 'jdbc:hive2://127.0.0.1:10009/?kyuubi.engineEnv.FLINK_HOME=/opt/flink;kyuubi.ha.namespace=kyuubi-ns-c;kyuubi.engine.type=FLINK_SQL'`
This method is extremely unfriendly to users.

I implemented a SessionConfAdvisor implementation class `org.apache.kyuubi.session.FileSessionConfAdvisor` based on the conf file that to manage session level configurations, so that we can access this cluster only by specifying a unique engine name, for example:
access cluster-a:
`beeline -u 'jdbc:hive2://127.0.0.1:10009/?kyuubi.session.conf.profile=cluster-a'`

access cluster-c:
`beeline -u 'jdbc:hive2://127.0.0.1:10009/?kyuubi.session.conf.profile=cluster-c'`

`kyuubi-session-cluster-a.conf` configuration is as follows:
```
    kyuubi.ha.namespace kyuubi-ns-a
    kyuubi.engine.type SPARK_SQL
    kyuubi.engine.pool.balance.policy POLLING

    kyuubi.engineEnv.SPARK_HOME /opt/spark30
    kyuubi.engineEnv.HADOOP_CONF_DIR /opt/hadoop_conf_dir
```

`kyuubi-session-cluster-c.conf` configuration is as follows:
```
    kyuubi.ha.namespace kyuubi-ns-a
    kyuubi.engine.type SPARK_SQL

    kyuubi.engineEnv.SPARK_HOME /opt/spark32
    kyuubi.engineEnv.HADOOP_CONF_DIR /opt/hadoop_conf_dir
```

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by turboFei** [comment]
It looks like `kyuubi.session.conf.tag` instead of `cluster`.

**Comment by ychris78** [comment]
> It looks like `kyuubi.session.conf.tag` instead of `cluster`.

Sorry for the trouble caused by the inappropriate title, which has been corrected

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala [hunk 1]
```diff
@@ -71,7 +70,7 @@ class FlinkProcessBuilder(
     }
 
     buffer += "-cp"
-    val classpathEntries = new util.LinkedHashSet[String]
+    val classpathEntries = new java.util.LinkedHashSet[String]
```
Comment (ychris78) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala:73]: The following error occurs during unit test, so the specified util package is displayed
image

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilderSuite.scala [hunk 2]
```diff
@@ -61,7 +60,7 @@ class FlinkProcessBuilderSuite extends KyuubiFunSuite {
   }
 
   private def constructClasspathStr(builder: FlinkProcessBuilder) = {
-    val classpathEntries = new util.LinkedHashSet[String]
+    val classpathEntries = new java.util.LinkedHashSet[String]
```
Comment (ychris78) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilderSuite.scala:63]: ditto

File: extensions/server/kyuubi-server-plugin/src/main/java/org/apache/kyuubi/plugin/FileSessionConfAdvisor.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+package org.apache.kyuubi.plugin;
+
+import java.io.*;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.kyuubi.Utils;
+import org.apache.kyuubi.config.KyuubiConf;
+
+public class FileSessionConfAdvisor implements SessionConfAdvisor {
```
Comment (pan3793) [inline extensions/server/kyuubi-server-plugin/src/main/java/org/apache/kyuubi/plugin/FileSessionConfAdvisor.java:26]: we can move class this to the server module

File: extensions/server/kyuubi-server-plugin/src/main/java/org/apache/kyuubi/plugin/SessionConfAdvisor.java [hunk 4]
```diff
@@ -24,7 +24,8 @@
 public interface SessionConfAdvisor {
   /** The returned conf will overwrite the session conf */
   @SuppressWarnings("unchecked")
-  default Map<String, String> getConfOverlay(String user, Map<String, String> sessionConf) {
+  default Map<String, String> getConfOverlay(String user, Map<String, String> sessionConf)
+      throws Exception {
```
Comment (pan3793) [inline extensions/server/kyuubi-server-plugin/src/main/java/org/apache/kyuubi/plugin/SessionConfAdvisor.java:28]: please avoid using checked exception as much as possible

File: extensions/server/kyuubi-server-plugin/src/main/java/org/apache/kyuubi/plugin/FileSessionConfAdvisor.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    if (sessionProfile == null) {
+      return sessionConf;
+    }
+    String pathName = "conf/kyuubi-session-" + sessionProfile + ".conf";
+    File propsFile = new File(pathName);
+    if (!propsFile.exists()) {
+      throw new FileNotFoundException("File: " + pathName + " is not found!");
+    }
+    scala.collection.immutable.Map<String, String> propertiesFromFile =
+        Utils.getPropertiesFromFile(propsFile);
```
Comment (pan3793) [inline extensions/server/kyuubi-server-plugin/src/main/java/org/apache/kyuubi/plugin/FileSessionConfAdvisor.java:41]: Use a `KyuubiConf` to load this file

File: extensions/server/kyuubi-server-plugin/pom.xml [hunk 6]
```diff
@@ -29,6 +29,13 @@
 
     <artifactId>kyuubi-server-plugin</artifactId>
     <name>Kyuubi Project Server Plugin</name>
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.kyuubi</groupId>
+            <artifactId>kyuubi-common_${scala.binary.version}</artifactId>
```
Comment (pan3793) [inline extensions/server/kyuubi-server-plugin/pom.xml:35]: please revert this change

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilderSuite.scala (hunk 2 above)
Comment (pan3793) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilderSuite.scala:63]: another option is `import java.util.{LinkedHashSet => JLinkedHashSet}`

File: kyuubi-server/src/test/scala/org/apache/kyuubi/plugin/PluginLoaderSuite.scala [hunk 7]
```diff
@@ -37,6 +40,18 @@ class PluginLoaderSuite extends KyuubiFunSuite {
...
+
+  test("test FileSessionConfAdvisor") {
+    val conf = new KyuubiConf(false)
+    conf.set(KyuubiConf.SESSION_CONF_ADVISOR, classOf[FileSessionConfAdvisor].getName)
+    conf.set(KyuubiConf.SESSION_PROFILE, "non.exists")
+    val msg3 = intercept[KyuubiException] {
+      val advisor = PluginLoader.loadSessionConfAdvisor(conf)
+      advisor.getConfOverlay("chris", conf.getAll.asJava)
+    }.getMessage
+    assert(msg3.startsWith("Failed when loading Kyuubi properties"))
```
Comment (pan3793) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/plugin/PluginLoaderSuite.scala:54]: please provide another positive test case

File: extensions/server/kyuubi-server-plugin/pom.xml (hunk 6 above)
Comment (ychris78) [reply extensions/server/kyuubi-server-plugin/pom.xml]: done

File: kyuubi-server/src/main/java/org/apache/kyuubi/session/FileSessionConfAdvisor.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.apache.kyuubi.Utils;
+import org.apache.kyuubi.config.KyuubiConf;
+import org.apache.kyuubi.plugin.SessionConfAdvisor;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+
+public class FileSessionConfAdvisor implements SessionConfAdvisor {
```
Comment (pan3793) [inline kyuubi-server/src/main/java/org/apache/kyuubi/session/FileSessionConfAdvisor.java:28]: better to write in scala to keep consistent

Comment (ychris78) [reply kyuubi-server/src/main/java/org/apache/kyuubi/session/FileSessionConfAdvisor.java]: done

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  override def getConfOverlay(
+      user: String,
+      sessionConf: util.Map[String, String]): util.Map[String, String] = {
+    val sessionProfile: String = sessionConf.get(KyuubiConf.SESSION_CONF_PROFILE.key)
+    if (sessionProfile == null) {
+      return Collections.emptyMap()
+    }
+    val pathName: String = s"conf/kyuubi-session-${sessionProfile}.conf"
+    val propsFile: File = new File(pathName)
+    Utils.getPropertiesFromFile(Option(propsFile)).asJava
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:41]: shall we use LocalCache?

otherwise, we need read the properties file every time.

Comment (ychris78) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:41]: As a session level plugin, it will only be read once when initializing the kyuubi session.

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:40]: read from disk on creating every session is a burden, let's use the cache here. on the other hand, it does not support hot load then.

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:40]: do you mean `getDefaultPropertiesFile` ? `getPropertiesFromFile` does not respect kyuubi home

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 10]
```diff
@@ -993,6 +993,16 @@ object KyuubiConf {
       .stringConf
       .createOptional
 
+  val SESSION_CONF_PROFILE: OptionalConfigEntry[String] =
+    buildConf("kyuubi.session.conf.profile")
+      .doc("When kyuubi.session.conf.advisor=org.apache.kyuubi.session.FileSessionConfAdvisor," +
```
[the remark this answers is not in the document]
Comment (ychris78) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: done

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 11]
```diff
@@ -993,6 +993,16 @@ object KyuubiConf {
...
+        "Specify a session level configuration file, which will be combined with default.conf " +
+        "to have an impact. The corresponding configuration file is " +
+        "conf/kyuubi-session-<profile>.conf")
```
[the remark this answers is not in the document]
Comment (ychris78) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: done

File: kyuubi-server/src/test/scala/org/apache/kyuubi/plugin/PluginLoaderSuite.scala (hunk 7 above)
Comment (ychris78) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/plugin/PluginLoaderSuite.scala]: done

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala (hunk 9 above)
Comment (ychris78) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala]: > shall we use LocalCache?
> 
> otherwise, we need read the properties file every time.

LocalCache has been added

Comment (ychris78) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala]: > do you mean `getDefaultPropertiesFile` ? `getPropertiesFromFile` does not respect kyuubi home
thanks, fixed

**Comment by turboFei** [comment]
Could you use guava local cache？

**Comment by ychris78** [comment]
> 

done

**Comment by ychris78** [comment]
> Could you use guava local cache？

done

**Review by pan3793 — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 12]
```diff
@@ -993,6 +993,17 @@ object KyuubiConf {
       .stringConf
       .createOptional
 
+  val SESSION_CONF_PROFILE: OptionalConfigEntry[String] =
+    buildConf("kyuubi.session.conf.profile")
+      .doc("Specify a profile to load session-level configurations from " +
+        "`$KYUUBI_CONF_DIR/kyuubi-session-<profile>.conf`. " +
+        "If the file does not exist, error log will be printed. " +
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1000]:
```suggestion
        "This configuration will be ignored if the file does not exist. " +
```

File: kyuubi-server/src/test/resources/kyuubi-session-cluster-a.conf [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+#
+
+kyuubi.ha.namespace kyuubi-ns-a
+kyuubi.engine.type SPARK_SQL
+kyuubi.engine.pool.balance.policy POLLING
+
+kyuubi.engineEnv.SPARK_HOME "/opt/spark30"
+kyuubi.engineEnv.HADOOP_CONF_DIR  "/opt/hadoop_conf_dir"
+
+
```
Comment (pan3793) [inline kyuubi-server/src/test/resources/kyuubi-session-cluster-a.conf:25]:
```suggestion
kyuubi.engineEnv.SPARK_HOME /opt/spark30
kyuubi.engineEnv.HADOOP_CONF_DIR /opt/hadoop_conf_dir
```

File: docs/deployment/settings.md [hunk 14]
```diff
@@ -462,6 +462,7 @@ Key | Default | Meaning | Type | Since
...
 kyuubi.session.conf.advisor|&lt;undefined&gt;|A config advisor plugin for Kyuubi Server. This plugin can provide some custom configs for different user or session configs and overwrite the session configs before open a new session. This config value should be a class which is a child of 'org.apache.kyuubi.plugin.SessionConfAdvisor' which has zero-arg constructor.|string|1.5.0
 kyuubi.session.conf.ignore.list||A comma separated list of ignored keys. If the client connection contains any of them, the key and the corresponding value will be removed silently during engine bootstrap and connection setup. Note that this rule is for server-side protection defined via administrators to prevent some essential configs from tampering but will not forbid users to set dynamic configurations via SET syntax.|seq|1.2.0
+kyuubi.session.conf.profile|&lt;undefined&gt;|Specify a profile to load session-level configurations from `$KYUUBI_CONF_DIR/kyuubi-session-<profile>.conf`. If the file does not exist, error log will be printed. This configuration only has effect when `kyuubi.session.conf.advisor` is set as `org.apache.kyuubi.session.FileSessionConfAdvisor`.|string|1.7.0
```
Comment (pan3793) [inline docs/deployment/settings.md:465]:
```suggestion
kyuubi.session.conf.profile|&lt;undefined&gt;|Specify a profile to load session-level configurations from `$KYUUBI_CONF_DIR/kyuubi-session-<profile>.conf`. This configuration will be ignored if the file does not exist. This configuration only has effect when `kyuubi.session.conf.advisor` is set as `org.apache.kyuubi.session.FileSessionConfAdvisor`.|string|1.7.0
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+      sessionConf: util.Map[String, String]): util.Map[String, String] = {
+    val profile: String = sessionConf.get(KyuubiConf.SESSION_CONF_PROFILE.key)
+    profile match {
+      case null => Collections.emptyMap()
+      case _ =>
+        sessionConfCache.get(
+          profile,
+          new Callable[util.Map[String, String]]() {
+            override def call(): util.Map[String, String] = {
+              val propsFile = Utils.getPropertiesFile(s"kyuubi-session-${profile}.conf")
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:45]:
```suggestion
              val propsFile = Utils.getPropertiesFile(s"kyuubi-session-$profile.conf")
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+              }
+            }
+          })
+    }
+  }
+}
+
+object FileSessionConfAdvisor {
+  private val sessionConfCache: Cache[String, util.Map[String, String]] =
+    CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.DAYS).build()
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:63]: How about adding a config for expire time？
Just a static string and a default value. And add description into the plugin config item.

**Review by turboFei — CHANGES_REQUESTED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala (hunk 16 above)
Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:63]: The default expire time is too long.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+          new Callable[util.Map[String, String]]() {
+            override def call(): util.Map[String, String] = {
+              val propsFile = Utils.getPropertiesFile(s"kyuubi-session-$profile.conf")
+              propsFile match {
+                case None =>
+                  error("File not found: $KYUUBI_CONF_DIR/" + s"kyuubi-session-$profile.conf")
+                  Collections.emptyMap()
+                case Some(_) =>
+                  val conf = Utils.getPropertiesFromFile(propsFile).asJava
+                  sessionConfCache.put(profile, conf)
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:52]: no need to do cache manually, guava will help you cache the returned value

Comment (ychris78) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala]: thanks, removed

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala (hunk 16 above)
Comment (ychris78) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala]: changed to 10 mins

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+                  Utils.getPropertiesFromFile(propsFile).asJava
+              }
+            }
+          })
+    }
+  }
+}
+
+object FileSessionConfAdvisor {
+  private val sessionConfCache: Cache[String, util.Map[String, String]] =
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:60]: import java.util.{Map => JMap}

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:60]:
```
    CacheBuilder.newBuilder()
      .expireAfterWrite(...)
      .build(new CacheLoader[....] {
        override def load(...): ... = {
        }
      })
```

Comment (ychris78) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/FileSessionConfAdvisor.scala:60]: > 

Is there any disadvantage in loading through get()?

**Review by turboFei — APPROVED** [review]
LGTM

**Comment by turboFei** [comment]
Some confusion on the pr description:

>  and it is extremely unsafe to expose namespaces to users.

1. Customer can check it on spark UI, environments tab.
2. Customer can check it with sql command `set spark.kyuubi.ha.namespace`.

**Comment by ychris78** [comment]
> Some confusion on the pr description:
> 
> > and it is extremely unsafe to expose namespaces to users.
> 
> 1. Customer can check it on spark UI, environments tab.
> 2. Customer can check it with sql command `set spark.kyuubi.ha.namespace`.

removed

**Review by cxzl25 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks @ychris78 and all reviewers, merging to master

**Comment by ychris78** [comment]
> Thanks @ychris78 and all reviewers, merging to master

thanks, merged to master
